zkevm-circuits/src/evm_circuit/util/math_gadget.rs

```diff
@@ -816,3 +817,316 @@ impl<F: FieldExt, const NUM_BYTES: usize> MaxGadget<F, NUM_BYTES> {
         Ok(select::value(lt, rhs, lhs))
     }
 }
+
+fn generate_polynomial<F:FieldExt>(now_cell: Cell<F>, now_num: u64, lim: u64) -> Expression<F> {
+    let mut now_expression = Expression::Constant(F::from_u128(1u128));
+    let mut now_invert_expression = now_expression.clone();
+    for not_equal_num in 0..lim {
+        if not_equal_num != now_num {
+            now_expression = now_expression * (now_cell.expr() - not_equal_num.expr());
+            let invert = if now_num < not_equal_num {
+                -F::from_u128((not_equal_num.clone() - now_num) as u128).invert().unwrap_or(F::zero())
+            } else {
+                F::from_u128((now_num - not_equal_num.clone()) as u128).invert().unwrap_or(F::zero())
+            };
+            now_invert_expression = now_invert_expression * invert;
+        }
+    }
+    now_expression * now_invert_expression
+}
```

Comment:
In fact, this function is generating Lagrange base polynomial. So rename it will be more clear.

Some other coding style suggestions: (1) we should name function parameter focusing on itself(caller) rather than who calls them(caller). So names like now_cell/now_num is not very good naming. (2) "unwrap_or(F::zero())" should never ever happens. We'd better use unwrap() directly. Don't hide bugs, let them crash so we can find and fix them. (3) now_invert_expression should be F rather than expression, calculate them eagerly, not lazily (4) lim -> limit, somethings lim reminds reader of limb, (used in wrong field ECC), it is confusing. 

here is my version:

```
fn generate_lagrange_base_polynomial<F:FieldExt>(cell: Cell<F>, idx: u64, domain_size: u64) -> Expression<F> {
    let mut base_ploy = Expression::Constant(F::from_u128(1u128));
    let mut accumulated_inverse = F::from_u128(1u128);
    for x in 0..domain_size {
        if x != idx {
            base_ploy = base_ploy * (now_cell.expr() - idx.expr());
            let inverse = if idx < x {
                -F::from_u128((x - idx) as u128).invert().unwrap()
            } else {
                F::from_u128((idx - x) as u128).invert().unwrap();
            };
            accumulated_inverse = accumulated_inverse * inverse;
        }
    }
    base_ploy * accumulated_inverse.exp()
```